### DISCUSSION TITLE
Inconsistent update with template x-if

### DISCUSSION THREAD

**Discussion by cjloong** [alpinejs/alpine#3286] [7. Bugs (archived)]
### Discussed in https://github.com/alpinejs/alpine/discussions/2802

<div type='discussions-op-text'>

<sup>Originally posted by **cjloong** April  1, 2022</sup>
Hi there,
I have a page structured as follows:
section1 (x-if="bar")
- input x-model="foo"

section2
- part 1 (x-if="!bar")
- part 2 (appears in section 1 and 2)
- part 3 (x-if="!bar")

The code to control the above is quite complicated. We discovered an intermittent issue where section 2, part 1 sometimes does not appear. Same as part 3.

I have manage to condense the issue into a codepen as follows:
https://codepen.io/cjloong/pen/zYpzQGr

Click on the "Simulation Error" button to reproduce the error.

</div>

**Comment by cjloong** [comment]
Hi, anyone have any idea on this. Its reproducible.

**Comment by Foxbud** [comment]
I believe I'm experiencing a related issue with `x-show`. It almost seems like `x-show` breaks when its dependent value toggles too quickly.

In this self-contained example, clicking `Enable` under the `Unexpected Behavior` header results in both the `Enable` and `Disable` buttons being displayed despite them having mutually exclusive display conditions. Then the `Expected Behavior` header demonstrates how adding a slight delay seems to fix the issue.

```html
<html>
	<head>
		<script src="https://cdnjs.cloudflare.com/ajax/libs/alpinejs/3.10.2/cdn.min.js" defer></script>
	</head>
	<body>
		<h1>Unexpected Behavior</h1>
		<div x-data="{state: false}"
			 x-init="$watch('state', val => {
						if (val) {state = false}
					})">
			<button type="button"
					x-show="state"
					@click="state = false">
				Disable
			</button>
			<button type="button"
					x-show="!state"
					@click="state = true">
				Enable
			</button>
		</div>
		
		<h1>Expected Behavior</h1>
		<div x-data="{state: false}"
			 x-init="$watch('state', val => {
						setTimeout(() => {if (val) {state = false}}, 100)
					})">
			<button type="button"
					x-show="state"
					@click="state = false">
				Disable
			</button>
			<button type="button"
					x-show="!state"
					@click="state = true">
				Enable
			</button>
		</div>
	</body>
</html>
```
Codepen: https://codepen.io/foxbud/pen/xxYRjQv

**Comment by vlad-iliescu-tallarium** [comment]
I've had a similar issue with `x-if`, which I think is the cause for the behavior seen in the codepen. When `x-if` hides the elements under it, it also removes the effects of its children ([ref](https://github.com/alpinejs/alpine/blob/9116006aa48c39709451eea790e7328e9a12f295/packages/alpinejs/src/directives/x-if.js#L30)), the issue is that this happens mid loop, while the job queue is being processed ([loop in question](https://github.com/alpinejs/alpine/blob/9116006aa48c39709451eea790e7328e9a12f295/packages/alpinejs/src/scheduler.js#L32)).
On button toggle (`this.foo = ""; this.bar = null;`) the jobs in the queue should be 0: "Update input foo due to `x-model`" 1: "Hide Screen1 due to `x-if`", 2: "Show Screen2:PART1 due to `x-if`" 3: "Show Screen2:PART2 due to `x-if`".
What will happen here is that when we "Hide Screen1", we will remove item 0 from the queue, but the index stays the same in the loop. So after processing queue[1], i will become 2, but since the queue was mutated, it will jump directly to "Show Screen2:PART2".
In the coderpen example, changing the toggle loop to:
```
         if (this.toggle) {
            this.bar = null;
            this.foo = "";
          } else {
            this.foo = "some text for input field";
            this.bar = { hello: "world" };
          }
```
will solve the issue, as the queue will become 0: "Hide Screen1 due to `x-if`", 1: "Show Screen2:PART1 due to `x-if`" 2: "Show Screen2:PART2 due to `x-if`" 3: "Update input foo due to `x-model`" and removing the item from the end won't make a difference.

This probably requires a proper bugfix, though.

**Comment by mgschoen** [comment]
Adding a real-world example here to illustrate practical implications of this bug. I'm building an audio player with Alpine.js. I'm using a store that represents playback state of an HTML5 audio element. The store listens to a couple of [HTMLMediaElement events](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement#events) such as `loadstart` and `playing` to determine if the audio is currently loading, playing or paused. It looks something like this:

```
const audio = document.createElement('audio');

export const WebAudioStore = {
    isPlaying: false,
    isLoading: false,

    init() {
        audio.addEventListener('loadstart', () => this.onStateChanged());
        audio.addEventListener('playing', () => this.onStateChanged());
        ...
    },

    onStateChanged() {
        // set isPlaying and isLoading depending on audio element state
    }
}
```

The bug occurs when a user jumps to another position in the audio that is already buffered. In that case, the audio element fires `loadstart`, directly followed by `playing`, causing state to switch from `isLoading=true` to `isPlaying=true` within milliseconds. Due to the bug described above, the UI shows both loading and playing state afterwards.

To anyone interested, I solved the issue by adding a timeout. This cannot be a permanent solution though since it slows down responsivity of the UI. My solution looks something like this:

```
let isStateUpdatePrevented = false;
let isStateUpdatePending = false;

export const WebAudioStore = {
    ...
    onStateChanged() {
        if (isStateUpdatePrevented) {
            isStateUpdatePending = true;
            return;
        }

        // set isPlaying and isLoading depending on audio element state

        isStateUpdatePending = false;
        isStateUpdatePrevented = true;
        setTimeout(() => this.onStateUpdateTimeout(), 25);
    },

    onStateUpdateTimeout() {
        isStateUpdatePrevented = false;
        if (isStateUpdatePending) {
            this.onStateChanged();
        }
    }
}
```

**Comment by josh-tt** [comment]
Has this issue also been addressed by the x-show fix?

**Comment by mgschoen** [comment]
@josh-tt Unfortunately not, the x-if issue still exists :(

**Comment by josh-tt** [comment]
Bummer, thanks. I've been on an older version of Alpine and just ran into issues with templates and stores. 3.8.1 working ok for me (some memory issues when fast clicking, but working ok), but anything above that is behaving inconsistently.

**Comment by dfallman** [comment]
We've had a similar problem with x-if, see #3003 creating a pretty wild memory leak

Our solution was to use $nextTick()

**Comment by bakman2** [comment]
I encounter issues with template swapping and wonder if this is the same issue or am I just missing something with this:

```
<body x-data="{test:['hello','bye'], show:true}">
    <template x-for="e in test">

        <template x-if="show">
            <div x-text="e"></div>
        </template>

        <template x-if="!show">
            <div>show is false</div>
        </template>

    </template>
</body>

```

If I manually set `show:false` the second template is not shown. 

arg - ignore, i had to wrap the templates in a div (note to self: template can only a single root node)

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/alpinejs/alpine/discussions/1423
